R/lint.R

```diff
@@ -100,32 +90,17 @@ lint <- function(filename, linters = NULL, cache = FALSE, ..., parse_settings =
     cache <- FALSE
   }
 
-  for (expr in source_expressions$expressions) {
-    for (linter in names(linters)) {
-      expr_lints <- NULL
-      if (isTRUE(cache) && has_lint(lint_cache, expr, linter)) {
-        # retrieve_lint() might return NULL if missing line number is encountered.
-        # It could be caused by nolint comments.
-        expr_lints <- retrieve_lint(lint_cache, expr, linter, source_expressions$lines)
-      }
-
-      if (is.null(expr_lints)) {
-        expr_lints <- flatten_lints(linters[[linter]](expr))
-
-        for (i in seq_along(expr_lints)) {
-          expr_lints[[i]]$linter <- linter
-        }
-
-        if (isTRUE(cache)) {
-          cache_lint(lint_cache, expr, linter, expr_lints)
-        }
-      }
-      lints[[itr <- itr + 1L]] <- expr_lints
-    }
-  }
+  n_linters <- length(linters)
+  n_checks <- length(source_expressions$expressions) * n_linters
+  lints <- lapply(seq_len(n_checks), function(check_i) {
+    # modular arithmetic to vary linters first, then expressions
+    expr <- source_expressions$expressions[[((check_i - 1L) %/% n_linters) + 1L]]
```

Comment:
not the prettiest -- two other approaches:

 1. Use `DF <- expand.grid(seq_len(n_linters), seq_len(n_expr))`, then `DF[check_i, ]` has the right coords
 2. Use `M <- matrix(nrow = n_linters, ncol = n_expr)`, then `col(M)` and `row(M)` have the right coords

Reply:
I think we're sacrificing readability for cyclomatic complexity here.
NB: did you take a look at @russHyde's PR for caching as a class?

Reply:
Looks like a more systematic way of accomplishing what's attempted in this PR. Nice! Any hope for progress @russHyde ? :grimacing: